src/core2cuda.rkt

```diff
@@ -0,0 +1,152 @@
+#lang racket
+
+(require generic-flonum)
+(require "imperative.rkt")
+
+(provide cuda-header
+         cuda-supported
+         cuda-reserved
+         core->cuda
+         type->cuda
+         params->cuda      ; required by core2h.rkt
+         set-cuda-header!
+         set-unknown->cuda!)
+
+(define cuda-header (const "#include <fenv.h>\n#include <math.h>\n#include <stdint.h>\n__device__ double e = 2.71828182845904523536;\n\n"))
+(define cuda-supported
+  (supported-list
+    (invert-op-proc (curry set-member? '(array dim size ref for for* tensor tensor*)))
+    fpcore-consts
+    (curry set-member? '(binary32 binary64 binary80 integer))
+    (invert-rnd-mode-proc (curry equal? 'nearestAway))
+    #f))
+
+(define (default-unknown->cuda ctx op args)
+  (define type (type->cuda (ctx-lookup-prop ctx ':precision)))
+  (format "~a~a(~a)"
+          op
+          (cuda-type->suffix type)
+          (string-join args ", ")))
+
+(define unknown->cuda (make-parameter default-unknown->cuda))
+
+(define (set-cuda-header! gen-proc)
+  (set! cuda-header (gen-proc cuda-header)))
+
+(define (set-unknown->cuda! gen-proc)
+  (unknown->cuda (gen-proc (unknown->cuda))))
+
+(define cuda-reserved  ; Language-specific reserved names (avoid name collisions)
+  '(auto break case char const continue default
+    do double else enum extern float for goto if
+    inline int long register restrict return
+    short signed sizeof static struct switch
+    typedef union unsigned void volatile while))
+
+(define/match (cuda-type->suffix type)
+  [("int64_t") ""]
+  [("double") ""]
+  [("float") "f"]
+  [("long double") "l"])
```

Comment:
I don't think `long double` is used on the device side.

Reply:
Just checked, also not used on device side, removed this too.